ratis-shell/src/main/java/org/apache/ratis/shell/cli/RaftUtils.java

```diff
@@ -86,4 +103,111 @@ public static RaftClient createClient(RaftGroup raftGroup) {
+    for (T t : list) {
+      try {
+        K ret = function.apply(t);
+        if (ret != null) {
+          return ret;
+        }
+      } catch (Throwable e) {
+        e.printStackTrace();
+      }
+    }
+    return null;
+  }
+
+
+  public static List<RaftPeer> buildRaftPeersFromStr(String peers) {
+    List<InetSocketAddress> addresses = new ArrayList<>();
+    String[] peersArray = peers.split(",");
+    for (String peer : peersArray) {
+      addresses.add(parseInetSocketAddress(peer));
+    }
+
+    return addresses.stream()
+        .map(addr -> RaftPeer.newBuilder()
+            .setId(RaftUtils.getPeerId(addr))
+            .setAddress(addr)
+            .build()
+        ).collect(Collectors.toList());
+  }
+
+  public static RaftGroupId buildRaftGroupIdFromStr(String groupId) {
+    return (groupId != null && !groupId.equals("")) ? RaftGroupId.valueOf(UUID.fromString(groupId))
+        : DEFAULT_RAFT_GROUP_ID;
+  }
+
+  public static RaftGroupId retrieveRemoteGroupId(RaftGroupId raftGroupIdFromConfig,
+                                                  List<RaftPeer> peers,
+                                                  RaftClient client, PrintStream printStream) throws IOException {
+    RaftGroupId remoteGroupId;
+    if (raftGroupIdFromConfig != DEFAULT_RAFT_GROUP_ID) {
+      return raftGroupIdFromConfig;
+    } else {
+      final List<RaftGroupId> groupIds = runFunction(peers,
+          p -> client.getGroupManagementApi((p.getId())).list().getGroupIds());
+
+      if (groupIds == null) {
+        printStream.println("Failed to get group ID from " + peers);
+        throw new IOException("Failed to get group ID from " + peers);
+      } else if (groupIds.size() == 1) {
+        remoteGroupId = groupIds.get(0);
+      } else {
+        printStream.println("There are more than one groups, you should specific one. " + groupIds);
+        throw new IOException("There are more than one groups, you should specific one. " + groupIds);
+      }
+    }
+
+    return remoteGroupId;
+  }
+
+  public static GroupInfoReply retrieveGroupInfoByGroupId(RaftGroupId remoteGroupId, List<RaftPeer> peers,
+                                                          RaftClient client, PrintStream printStream)
+      throws IOException {
+    GroupInfoReply groupInfoReply = runFunction(peers,
+        p -> client.getGroupManagementApi((p.getId())).info(remoteGroupId));
+    processReply(groupInfoReply,
+        printStream::println, "Failed to get group info for group id " + remoteGroupId.getUuid() + " from " + peers);
+    return groupInfoReply;
+  }
+
+  public static void processReply(RaftClientReply reply, Consumer<String> printer, String message) throws IOException {
+    processReplyInternal(reply, () -> printer.accept(message));
+  }
+
+  private static void processReplyInternal(RaftClientReply reply, Runnable printer) throws IOException {
```

Comment:
Combine this two methods and use ` Supplier<String>` for  `message`, i.e.
```java
  public static void processReply(RaftClientReply reply, Consumer<String> printer, Supplier<String> message)
      throws IOException {
    if (reply == null || !reply.isSuccess()) {
      final RaftException e = Optional.ofNullable(reply)
          .map(RaftClientReply::getException)
          .orElseGet(() -> new RaftException("Reply: " + reply));
      printer.accept(message.get());
      throw new IOException(e.getMessage(), e);
    }
  }
```

Reply:
Yeah sure!